File: php/menus.php
<?php

require_once 'entity.php';
require_once './../php_aux/menu_util.php';
require_once 'config.php';

class Category extends Entity
{
    public static $resource = '/menus/';
    public static $json_name = 'menu';

    public function __construct()
    {
        $this->json_property('id','integer');
        $this->json_property('name','string');
        $this->json_property('menu_handle','string');
        $this->json_property('menu_type','integer');
        $this->json_property('menu_items', 'MenuItem', $many = True,
                             $recursive = True);
    }

    public function  menu_items_in_order(){
        #Return a list of menu_items sorted by their position
        $return_array = $this->menu_items;
        usort($return_array, function($a, $b){
            return $a->position <=> $b->postition;
        })
        return $return_array;
    }
}

class MenuItem extends Entity
{
    public function __construct()
    {
      $this->json_property('id','integer');
      $this->json_property('name','string');
      $this->json_property('link_url','string');
      $this->json_property('link_type','integer');
      $this->json_property('position','integer');
    }

    public function url($domain = Null){
      /*Check to see if the menu type is a redirect or an
            internal link, then return a full url constructed from the
            link_uri and the domain
      */
      $server_name = SERVER_NAME;
      if($domain and $this->link_type == INTERNAL_PAGE){
          $fmt = "http://%s.%s/%s/";
          return sprintf($fmt, (string)$domain->sub_domain, $server_name,
          $this->link_uri);
      }
      return $this->link_uri;
    }
}

File: python/menus.py
import sdk.python.entities
import sdk.python.util.menu_util as menu_util
from sdk.python.entities import Property
from operator import attrgetter


class MenuItem(sdk.python.entities.Entity):

    resource = '/menu_items'
    json_name = 'menu_item'

    id = Property(int)
    name = Property(str)
    link_type = Property(int)
    link_uri = Property(str)
    position = Property(int)

    def url(self, host, domain=None):
        """ Check to see if the menu type is a redirect or an
            internal link, then return a full url constructed from the
            link_uri and the domain
        """
        if domain and self.link_type == menu_util.INTERNAL_PAGE:
            return 'http://{0}.{1}/{2}/'.format(str(domain.sub_domain),
                                                host,
                                                self.link_uri)
        return self.link_uri


class Menu(sdk.python.entities.Entity):

    resource = '/menus/'
    json_name = 'menu'

    id = Property(int)
    name = Property(str)
    menu_handle = Property(str)
    menu_type = Property(int)
    menu_items = Property(MenuItem)

    def menu_items_in_order(self):
        """ Return a list of menu_items sorted by their position """
        return sorted(self.menu_items, key=attrgetter('position'))
